### PR TITLE
Get rid of universal variable indices.

### PR DESCRIPTION
**Pull request by smimram**
It might help clarifying #1035.

### CODE REVIEW

**Comment by toots** [comment]
Oh my. Gonna have to catch up on the typing stuff...

Question: how do you account for things like: `list.hd(default='a, ['a]) -> 'a` where `'a` is the same on all 3 spots?

**Comment by smimram** [comment]
Like this: https://github.com/savonet/liquidsoap/pull/1037/files#diff-0c0cdd98d034cac51af8683f71f93cafL96

You declare your type variable before with a `let` and use it three times.

**Comment by toots** [comment]
> Like this: https://github.com/savonet/liquidsoap/pull/1037/files#diff-0c0cdd98d034cac51af8683f71f93cafL96
> 
> You declare your type variable before with a `let` and use it three times.

Awesome. This looks great, give me a second to review it and let's merge soon. One question we should answer before though: why were those made to be cached in the first place?

**Comment by smimram** [comment]
My guess is that they were really numbered first and then caching was introduced in order not to change the API. At least, I understand the old comment in this way.

File: src/conversions/audio_to_stereo.ml [hunk 1]
```diff
@@ -68,9 +68,7 @@ object
 end
 
 let () =
-  let input_kind =
-    Lang.kind_type_of_kind_format ~fresh:1 Lang.audio_variable
-  in
+  let input_kind = Lang.kind_type_of_kind_format Lang.audio_variable in
```
Comment (toots) [inline src/conversions/audio_to_stereo.ml]:
```suggestion
  let input_kind =
    Lang.kind_type_of_kind_format Lang.audio_variable
  in
```

File: src/lang/builtins_list.ml [hunk 2]
```diff
@@ -123,12 +124,13 @@ let () =
...
 let () =
+  let a = Lang.univ_t () in
   add_builtin "list.sort" ~cat:List
     ~descr:"Sort a list according to a comparison function."
     ["",
-     Lang.fun_t [false,"",Lang.univ_t 1;false,"",Lang.univ_t 1] Lang.int_t,
+     Lang.fun_t [false,"",a;false,"",a] Lang.int_t,
      None, None ;
-     "",Lang.list_t (Lang.univ_t 1),None,None] (Lang.list_t (Lang.univ_t 1))
+     "",Lang.list_t a,None,None] (Lang.list_t a)
```
Comment (toots) [inline src/lang/builtins_list.ml]: 🎉

File: src/lang/builtins_time.ml [hunk 3]
```diff
@@ -48,9 +48,10 @@ let () =
 let () =
   add_builtin ~cat:Sys "time" ~descr:"Return the current time since 00:00:00 GMT, Jan. 1, 1970, in seconds."
     [] Lang.float_t (fun _ ->
-      Lang.float (Unix.gettimeofday ()));
+        Lang.float (Unix.gettimeofday ()));
+  let univ = Lang.univ_t () in
```
Comment (toots) [inline src/lang/builtins_time.ml]:
```suggestion
  let t = Lang.univ_t () in
```

File: src/lang/builtins_time.ml [hunk 4]
```diff
@@ -48,9 +48,10 @@ let () =
...
   let execute cb tm =
-    Lang.apply cb ~t:(Lang.univ_t 1)
+    Lang.apply cb ~t:univ
```
Comment (toots) [inline src/lang/builtins_time.ml]:
```suggestion
    Lang.apply cb ~t
```

File: src/lang/lang.ml [hunk 5]
```diff
@@ -63,30 +63,11 @@ let variable_t = Term.variable_t
...
-          v
-let string_getter_t n = univ_t ~constraints:[T.Getter T.String] n
-let float_getter_t n = univ_t ~constraints:[T.Getter T.Float] n
-let int_getter_t n = univ_t ~constraints:[T.Getter T.Int] n
-let bool_getter_t n = univ_t ~constraints:[T.Getter T.Bool] n
+let univ_t ?(constraints=[]) () = T.fresh ~level:0 ~constraints ~pos:None
+let string_getter_t () = univ_t ~constraints:[T.Getter T.String] ()
+let float_getter_t () = univ_t ~constraints:[T.Getter T.Float] ()
+let int_getter_t () = univ_t ~constraints:[T.Getter T.Int] ()
+let bool_getter_t () = univ_t ~constraints:[T.Getter T.Bool] ()
```
Comment (toots) [inline src/lang/lang.ml]: giphy

**Pull request merged**